front/migrations/20240906_2_backfill_agents_groupIds.ts

```diff
@@ -0,0 +1,155 @@
+import { DataSourceViewResource } from "@app/lib/resources/data_source_view_resource";
+import { makeScript } from "@app/scripts/helpers";
+
+makeScript({}, async ({ execute }, logger) => {
+  // All workspaces that have at least one agent
+  const workspaceIds = await getDistinctWorkspaceIds();
+
+  const workspaceChunks = _.chunk(workspaceIds, 8);
+
+  for (const workspaceChunk of workspaceChunks) {
+    await Promise.all(
+      workspaceChunk.map((id) => updateAgentsForWorkspace(id, execute, logger))
+    );
+  }
+});
+
+async function getDistinctWorkspaceIds(): Promise<number[]> {
+  const workspaceIds = await AgentConfiguration.findAll({
+    attributes: [
+      [Sequelize.fn("DISTINCT", Sequelize.col("workspaceId")), "workspaceId"],
+    ],
+    raw: true,
+  });
+
+  return workspaceIds.map((entry) => entry.workspaceId);
+}
+
+async function updateAgentsForWorkspace(
+  workspaceId: number,
+  execute: boolean,
+  logger: Logger
+) {
+  const allAgents = await AgentConfiguration.findAll({
+    attributes: ["sId", "groupIds"],
+    where: { workspaceId, status: "active" },
+  });
+
+  // no need to update agents that already have groupIds
+  const agents = allAgents.filter(
+    (agent) => !agent.groupIds || agent.groupIds.length === 0
+  );
+
+  logger.info(
+    { workspaceId, count: agents.length },
+    "Updating agents for workspace"
+  );
+
+  const agentChunks = _.chunk(agents, 16);
+
+  // get workspace sid
+  const workspace = await Workspace.findByPk(workspaceId);
+
+  if (!workspace) {
+    logger.error(
+      {
+        workspaceId,
+      },
+      "Unexpected: Workspace not found"
+    );
+    return;
+  }
+
+  for (const agentChunk of agentChunks) {
+    const auth = await Authenticator.internalAdminForWorkspace(workspace.sId);
+
+    await Promise.all(
+      agentChunk.map((agent) => updateAgent(auth, agent, execute, logger))
+    );
+  }
+}
+
+async function updateAgent(
+  auth: Authenticator,
+  agent: AgentConfiguration,
+  execute: boolean,
+  logger: Logger
+) {
+  const assistant = await getAgentConfiguration(auth, agent.sId);
+  if (!assistant) {
+    logger.error(
+      {
+        agentId: agent.sId,
+      },
+      "Unexpected: Agent not found"
+    );
+    return;
+  }
+
+  const dataSourceViewIds = getDataSourceViewIdsFromActions(assistant.actions);
+
+  const groupIds = (
+    await DataSourceViewResource.fetchByIds(auth, dataSourceViewIds)
+  )
+    .map((view) => view.acl().aclEntries.map((entry) => entry.groupId))
+    .flat();
+
+  if (execute) {
+    await AgentConfiguration.update(
+      { groupIds },
+      { where: { sId: agent.sId } }
+    );
+    logger.info(
+      {
+        agentId: agent.sId,
+        execute,
+      },
+      "Updated agent"
+    );
+  } else {
+    logger.info(
+      {
+        agentId: agent.sId,
+        execute,
+      },
+      "Would have updated agent"
+    );
+  }
+}
+
+function getDataSourceViewIdsFromActions(
```

Comment:
Why not re-using this one?

https://github.com/dust-tt/dust/blob/c39e02f06cdc9dc99176c442e7d76b6d7452055a/front/pages/api/w/%5BwId%5D/assistant/agent_configurations/index.ts#L449

Reply:
because it's good for it to be private :+1:

Reply:
Not a big fan as this is pretty easy to deviate.

Reply:
As per IRL, will update to public during migration then optionally move back to private with a ts-expect-error